torch/csrc/autograd/function.cpp

```diff
@@ -36,95 +36,60 @@ AnomalyMetadata* Function::metadata() noexcept {
+                     std::vector<std::shared_ptr<Function>>& stack) {
+  func->release_variables();
+  auto saved_vars = func->saved_variables();
+  if (saved_vars) {
+    for (auto& saved_var : *saved_vars) {
+      saved_var.reset_grad_function();
+    }
   }
 
-  size_t value() {
-    return deleteFunctionRecursionDepth;
+  for (auto& edge : func->next_edges()) {
+    if (edge.function.use_count() == 1) {
+      stack.emplace_back(std::move(edge.function));
+    } else {
+      edge.function.reset();
+    }
   }
-};
+}
 
 /*
- * Note that the custom deleter deletes in BFS style. Without using
- * the custom deleter, the computation graph is deleted in a DFS style.
- * The BFS deletion is valid (and safe) because if a shared_ptr<Function>
- * 's reference count hits 0, nothing else will access it.
- */
+  * Fix for #5534: prevent stack overflow on deletion of deep computation graph
+  *
+  * Sometimes one can end up with a very big computation graph of Functions
+  * and Edges. Each std::shared_ptr<Function> contains a list of Edge, and
+  * each Edge contains a std::shared_ptr<Function>. Deleting a
+  * std::shared_ptr<Function> can trigger the recursive deletion of other
+  * std::shared_ptr<Function>'s: this can stack overflow if the graph
+  * is deep enough. Here is an example of such a graph:
+  *
+  * shared_ptr<Function> -> Edge -> shared_ptr<Function> -> Edge -> ... -> shared_ptr<Function>
+  *
+  * The solution here is to detect when we are decrementing away the last
+  * reference to a Function, and when doing so to buffer up the Function's
+  * that will be recursively decremented.  We can then decrement (and free)
+  * the original Function without causing a recursive cascade, before
+  * draining the buffer applying the same behavior.  This is, in effect,
+  * converting recursion to a loop, using a heap buffer in place of the
+  * recursive call stack.
+  */
 void deleteFunction(Function* function) {
-  RecursionDepthCounter recursion_depth;
-
-  if (recursion_depth.value() > deleteFunctionMaxRecursionDepth) {
-    deleteFunctionQueue.push_back(function);
-    return;
-  }
-
+  // To avoid stack overflow on large computational graphs,
+  // we need to track reference decrementing and freeing
+  // on the heap.
+  function->release_variables();
+  std::vector<std::shared_ptr<Function>> stack;
+  gatherFunctions(function, stack);
   delete function;
 
-  if (deleteFunctionQueue.empty()) {
-    return;
-  }
-  if (recursion_depth.value() != deleteFunctionMaxRecursionDepth) {
-    AT_ERROR("Only one deleter per thread should be able to process "
-             "the delete queue. Please open an issue.");
-  }
-  while (!deleteFunctionQueue.empty()) {
-    auto queued_function = deleteFunctionQueue.front();
-    deleteFunctionQueue.pop_front();
-    delete queued_function;
+  while (!stack.empty()) {
+    auto func = std::move(stack.back());
+    stack.pop_back();
+    assert(func.use_count() == 1);
```

Comment:
`asserts` are compiled away. You have to use `AT_ASSERT` if you really want them to be there